sparkmagic/sparkmagic/controllerwidget/addendpointwidget.py

```diff
@@ -1,74 +1,92 @@
 # Copyright (c) 2015  aggftw@gmail.com
 # Distributed under the terms of the Modified BSD License.
-from sparkmagic.controllerwidget.abstractmenuwidget import AbstractMenuWidget
+from .abstractmenuwidget import AbstractMenuWidget
 from sparkmagic.livyclientlib.endpoint import Endpoint
 import sparkmagic.utils.constants as constants
-
+import sparkmagic.utils.configuration as conf
+import importlib
+from collections import defaultdict
 
 class AddEndpointWidget(AbstractMenuWidget):
 
     def __init__(self, spark_controller, ipywidget_factory, ipython_display, endpoints, endpoints_dropdown_widget,
                  refresh_method):
         # This is nested
         super(AddEndpointWidget, self).__init__(spark_controller, ipywidget_factory, ipython_display, True)
-
+      
         widget_width = "800px"
 
         self.endpoints = endpoints
         self.endpoints_dropdown_widget = endpoints_dropdown_widget
         self.refresh_method = refresh_method
 
-        self.address_widget = self.ipywidget_factory.get_text(
-            description='Address:',
-            value='http://example.com/livy',
-            width=widget_width
-        )
-        self.user_widget = self.ipywidget_factory.get_text(
-            description='Username:',
-            value='username',
-            width=widget_width
-        )
-        self.password_widget = self.ipywidget_factory.get_text(
-            description='Password:',
-            value='password',
-            width=widget_width
-        )
-        self.auth = self.ipywidget_factory.get_dropdown(
-            options={constants.AUTH_KERBEROS: constants.AUTH_KERBEROS, constants.AUTH_BASIC: constants.AUTH_BASIC,
-                     constants.NO_AUTH: constants.NO_AUTH},
+        #map auth class path string to the instance of the class. 
+        self.auth_instances = {}
+        for auth in conf.auths_supported().values(): 
+            module, class_name = (auth).rsplit('.', 1)
+            events_handler_module = importlib.import_module(module)
+            auth_class = getattr(events_handler_module, class_name)
+            self.auth_instances[auth] = auth_class()
+
+        self.auth_type = self.ipywidget_factory.get_dropdown(
+            options = conf.auths_supported(),
             description=u"Auth type:"
         )
 
+        #defaultdict maps keys to list of values. .values() should then just return all of the values in one list not nested arrays. 
+        # maps instance to list of instances widgets. Useful so we can turn off / on display for widgets on dropdown change 
+        self.authWidgets = defaultdict(set)
```

Comment:
Instead of keeping a map of instances to sets of widgets, consider leveraging the fact we already have a map of auth types to instances (in self.auth_instances), and each auth instance already keeps track of it's own widgets. We could change the auth classes to instantiate the widgets at instance creation time, then just return the same widgets in `get_widgets`. Then each instance tracks its own set of widgets.